### PR TITLE
Update contributing guidelines to discourage contributions to certain areas

### PR DESCRIPTION
**Pull request by chaas**
As a follow up to https://github.com/MaterializeInc/materialize/issues/19743, we want to be more prescriptive about what types of external contributions we'll encourage, and more explicit guidelines for picking up an issue.

Surfaces has decided to disallow external contributions to the coordinator for two reasons:
* The cost of educating and performing extra-thorough reviews is not worth the benefit-cost/risk tradeoff, given the complexity and importance of this code
* Major coordinator changes should only be done as part of a more focused, planned body of work, again given the nuances. One-off changes picked up at random increase the risk of missed problems.

Questions for reviewer:
* Any other suggestions for both the "areas that are best suited for external contributions" and "ask first" sections? I picked persist there arbitrarily, and can remove if folks disagree.

This is one of a series of follow up items from the above issue.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -35,3 +38,12 @@ When landing large or substantial changes, we want to make sure users are aware
 - All new features
 - All API changes
 - Large bug fixes
+
+### Where to contribute
+Some areas that are best suited for external contributions are:
```
Comment (benesch) [inline CONTRIBUTING.md:43]:
```suggestion
Some areas that are well suited for external contributions are:
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -18,6 +18,9 @@ Materialize is written entirely in Rust. Rust API documentation is hosted at
 
 Prospective code contributors might find the [good first issue tag](https://github.com/MaterializeInc/materialize/issues?q=is%3Aopen+is%3Aissue+label%3A%22D-good+first+issue%22) useful.
 
+It's a good idea to comment on any Github issue you're planning to pick up to let folks know. That way, you can get
+proactive advice and speedier reviews.
```
Comment (benesch) [inline CONTRIBUTING.md:22]:
```suggestion
It's a good idea to comment on any GitHub issue you're planning to pick up to
let folks know. That way, you can get proactive advice and speedier reviews.

If you start work on the issue before hearing back from a Materialize engineer,
there is a risk that we may not be able to accept your changes. The team may not
have bandwidth to review your changes, or the code in question may require
specialized knowledge to modify without introducing bugs. When in doubt, wait
for a Materialize engineer to respond before starting work!
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -35,3 +38,12 @@ When landing large or substantial changes, we want to make sure users are aware
 - All new features
 - All API changes
 - Large bug fixes
+
+### Where to contribute
+Some areas that are best suited for external contributions are:
+- Adding SQL functions
+- Adding pg_catalog tables to support more [tools and integrations](https://materialize.com/docs/integrations/)
+
+It's best to ask on the Github issue before contributing to these areas:
```
Comment (benesch) [inline CONTRIBUTING.md:47]:
```suggestion
Areas that are not well suited for external contributions:
```

**Review by benesch — APPROVED** [review]
Added some suggestions for being even more discouraging of contributions in certain areas.

File: CONTRIBUTING.md (hunk 3 above)
Comment (benesch) [reply CONTRIBUTING.md:47]: I can't imagine @danhhz will object to this, but tagging him for explicit confirmation anyway.

**Comment by benesch** [comment]
> Any other suggestions for both the "areas that are best suited for external contributions" and "ask first" sections? I picked persist there arbitrarily, and can remove if folks disagree.

Probably worth re-asking this question in #eng-announce and seeing if other teams have areas where they'd like to discourage external contributions.

**Comment by benesch** [comment]
Thanks very much for putting this together, @chaas!

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
As part of merging this, @chaas, I'd love to do a quick pass of the issues marked "good first issue." There are only 42, so relabeling them should be tractable. I propose:

  * We rename "D-good first issue" to "D-good first issue (INTERNAL)" or somesuch
  * We add a new label "good for external contributors"
  * We only accept the "good for external contributors" label for issues that we really would accept help with.

**Comment by chaas** [comment]
Sounds good @benesch !

> We only apply the "good for external contributors" label to issues that we really would accept help with.

I can take a quick pass thru the current "D-good first issue" tickets and add "good for external contributors" to a few that are actually well-suited.

**Comment by chaas** [comment]
Created new "good for external contributor" tag, added it to three issues that are a good fit, and updated this guide to refer to that tag.

**Review by benesch — APPROVED** [review]
This is great. Thanks, @chaas!

**Pull request merged**